aptos-move/framework/supra-framework/sources/pbo_delegation_pool.move

```diff
@@ -1601,10 +1615,184 @@ module supra_framework::pbo_delegation_pool {
+                let (active, inactive, pending_inactive) = get_stake(pool_address, delegator);
+
+                // Ensure that all stake to be locked is made `active`.
+                if (active < principle_stake) {
+                    // The amount to lock can be covered by reactivating some previously unlocked stake.
+                    // Only reactivate the required amount to avoid unnecessarily interfering with 
+                    // in-progress withdrawals.
+                    let amount_to_reactivate = principle_stake - active;
+
+                    // Ensure that we do not try to reactivate more than the available `pending_inactive` stake.
+                    // This should be enforced by functions within `authorized_reactivate_stake`, but checking
+                    // again here makes the correctness of this function easier to reason about.
+                    if (amount_to_reactivate > pending_inactive) {
+                        amount_to_reactivate = pending_inactive;
+                    };
+
+                    // Reactivate the required amount of `pending_inactive` stake first, if there is any.
+                    authorized_reactivate_stake(delegator, pool_address, amount_to_reactivate);
+                    let active_and_pending_inactive = active + pending_inactive;
+                    
+                    if (active_and_pending_inactive < principle_stake) {
+                        // Need to reactivate some of the `inactive` stake.
+                        let amount_to_withdraw = principle_stake - active_and_pending_inactive;
+
+                        // Ensure that we do not try to withdraw more stake than the `inactive` stake.
+                        if (amount_to_withdraw > inactive) {
+                            amount_to_withdraw = inactive;
+                        };
+
+                        if (amount_to_withdraw > 0) {
+                            // Withdraw the minimum required amount to the admin's address.
+                            admin_withdraw(
+                                multisig_admin,
+                                pool_address,
+                                delegator,
+                                amount_to_withdraw
+                            );
+                            // Then allocate it to the delegator again.
+                            fund_delegator_stake(multisig_admin, pool_address, delegator, amount_to_withdraw);
+                        }
+                    }                    
+                };
+                // else: The amount to lock can be covered by the currently `active` stake.
+
+                // Update the delegator's principle stake and record the details of the lockup event.
+                let principle_stake_table =
+                    &mut (borrow_global_mut<DelegationPool>(pool_address).principle_stake);
+                table::upsert(principle_stake_table, delegator, principle_stake);
```

Comment:
Do you really want to `upsert` with `principle_stake` or `min(principle_stake,active+amount_to_reactivate+amount_to_withdraw)`? If `principle_stake` is larger than the actual locking that happened, it would result in larger portion getting unlocked.

Reply:
If we use the original `principle_stake` and a user has managed to withdraw more of their allocation than they are supposed to have at this time then they should not be able to withdraw anything until the time when they are supposed to be able to withdraw the next portion, and that portion will match the terms of the legal contract that they signed. If we change the `principle_stake` then we are no longer abiding by the original terms.

Also, if we use the original `principle_stake` and we convince the investors to re-delegate the additional tokens that they withdrew then we will have successfully restored the contract according to the original terms.